### PR TITLE
Implement OpenLog peripheral driver

### PR DESCRIPTION
**Pull request by jacobmcnamee**
Try to detect OpenLog on UARTA/B at boot. If found, scan files until a unique name is found. Files are named according to <prefix><id>.bin, where <prefix> is a user-configurable string and <id> is auto-generated and unique.

Right now a new file will be started a) at boot, or b) when the prefix string is changed. There is also a function openlog_file_break() which will do the same. Eventually breaking the file (via function call or settings change) should probably be done in a separate background thread as it can potentially take seconds to find the next unique id.

### CODE REVIEW

**Comment by fnoble** [comment]
Reviewing now - just a couple minor cosmetic points so far
 1. We use `/* */` style comments and not `//` in our C code
 2. Try to keep the max line length around 80 chars, going a bit over occasionally is fine but keeping lines around that length helps readability for some editor setups (and its in the style guide)

I'll follow up with any more substantive comments

File: src/peripherals/openlog.c [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+static bool command_mode_enter(usart_dma_state *s, void *params); // enter/verify command mode
+static bool openlog_discover(usart_dma_state *s, void *params);
+static bool openlog_reset(usart_dma_state *s, void *params);
+static bool file_size_get(usart_dma_state *s, void *params);
+static bool file_append_set(usart_dma_state *s, void *params); // data mode entered on success
+
+static bool log_file_update();
+static bool file_prefix_setting_notify(struct setting *setting, const char *val);
+
+static bool claim(usart_dma_state *s)
```
Comment (fnoble) [inline src/peripherals/openlog.c]: Could be moved up to `usart.c` and renamed e.g. `usart_claim_retry()`

File: src/peripherals/openlog.c [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+      return true;
+    }
+
+    chThdSleepMilliseconds(CLAIM_FAIL_DELAY_ms);
+  } while (++retry < CLAIM_RETRY_COUNT);
+
+  return false;
+}
+
+static void release(usart_dma_state *s)
```
Comment (fnoble) [inline src/peripherals/openlog.c]: Does this function need to exist?

File: src/peripherals/openlog.c [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+{
+  usart_release(s);
+}
+
+static bool usart_send_str(usart_dma_state *s, const char *str)
```
Comment (fnoble) [inline src/peripherals/openlog.c]: Again this is probably more generally useful, could move up to `usart.c`

File: src/peripherals/openlog.c [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+  // null terminate
+  if (dest_index < dest_len) {
+    dest[dest_index++] = 0;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+static bool parse_int(s32 *output, const char *str)
```
Comment (fnoble) [inline src/peripherals/openlog.c]: what about using `atoi` or `strtoi` instead?

File: src/peripherals/openlog.c [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+  bool success = false;
+
+  do {
+      success = func(s, params);
+  } while (!success && (++retry < COMMAND_RETRY_COUNT));
+
+  return success;
+}
+
+static bool strs_concat(char *dest, u32 dest_len, const char **srcs, u32 srcs_count)
```
Comment (fnoble) [inline src/peripherals/openlog.c]: Use `strncat`? We have the full standard library, may as well use it :)

File: src/peripherals/usart.c [hunk 6]
```diff
@@ -221,8 +228,16 @@ void usarts_enable(u32 ftdi_baud, u32 uarta_baud, u32 uartb_baud, bool do_precon
   chBSemInit(&uartb_state.claimed, FALSE);
   uartb_state.configured = true;
 
-  all_uarts_enabled = true;
+  if (do_configure_hooks) {
```
Comment (fnoble) [inline src/peripherals/usart.c]: Its not immediately obvious why these hooks aren't together with the radio ones. Maybe unify them so they are using the claim mechanism in the same way? or at least add a comment saying why they are separated.

File: src/peripherals/openlog.c [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+/** Hook called when the UART is configured. Used to discover a connected OpenLog */
+void openlog_configure_hook(usart_dma_state *s, char* uart_name)
+{
+  if (openlog_usart != 0) {
+    // already discovered an OpenLog
+    return;
+  }
+
+  if (!claim(s))
+  {
```
Comment (fnoble) [inline src/peripherals/openlog.c]: `{` should be on same line as `if`

File: src/peripherals/openlog.c [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+    command_mode_enter(s, 0);
+    return false;
+  }
+}
+
+static bool log_file_update(usart_dma_state *s)
+{
+  if ((file_id < FILE_ID_START) || (file_id > FILE_ID_LAST))
+  {
+      return false;
```
Comment (fnoble) [inline src/peripherals/openlog.c]: indentation and `{`

File: src/peripherals/openlog.c [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+  }
+
+  // enter command mode
+  if (!command_execute(s, command_mode_enter, 0)) {
+    log_info("OpenLog - error entering command mode - resetting\n");
```
Comment (fnoble) [inline src/peripherals/openlog.c]: Should some of these be `log_warn`s instead?

File: src/peripherals/openlog.c [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+      break;
+    }
+    str_index++;
+  }
+
+  *output = (negative ? -1 : 1) * result;
+  return started;
+}
+
+static bool terminators_match(char c, const char **terminators, u32 *terminator_indexes, u32 terminators_count)
```
Comment (fnoble) [inline src/peripherals/openlog.c]: I think this and/or `usart_wait_for_response` could do with a longer comment explaining the terminator mechanism.

**Comment by fnoble** [comment]
This is great! I think a lot of people are going to find this really useful. Let me know when you have  addressed these comments and I'll do another pass.

File: src/peripherals/openlog.c (hunk 1 above)
Comment (jacobmcnamee) [inline src/peripherals/openlog.c]: added usart_claim_timeout() which accepts a semaphore timeout in systicks

File: src/peripherals/openlog.c (hunk 4 above)
Comment (jacobmcnamee) [inline src/peripherals/openlog.c]: They silently return zero on failure.

File: src/peripherals/usart.c (hunk 6 above)
Comment (jacobmcnamee) [inline src/peripherals/usart.c]: The radio "preconfigure" hooks are a bit of a hack right now. They must be called before (and only before) any of the normal uart framework is operational. The openlog hooks use the standard claim() / release() mechanism. I think it's clear from the *pre*configure naming.

**Pull request closed without merging**